mixins/provider-mixin.js

```diff
@@ -0,0 +1,38 @@
+export const ProviderMixin = superclass => class extends superclass {
+
+	constructor() {
+		super();
+		this._providers = new Map();
+		this.addEventListener('d2l-request-provider', this._handleRequest);
+	}
+
+	_handleRequest(e) {
+		if (this._providers.has(e.detail.key)) {
+			e.detail.provider = this._providers.get(e.detail.key);
+			e.stopPropagation();
+		}
+	}
+
+	provide(key, factory) {
```

Comment:
Could the second variable be called something like `obj` instead of `factory`? The implication with `factory` is that they're providing a factory that needs to be called in order to get the thing... which I think is just one possible interpretation. We could use this to provide strings or objects that have already been created and just have other methods on them.

Reply:
Yeh, the version that I originally created that Ryan copied this from was from a YouTube talk that Justin gave a couple of years ago, where he used the `factory` metaphor and then the requestor had to call the provided thing as a function. But after I'd implemented it I realized it didn't necessarily add a whole lot of value to make the thing provided always be a factor. In Justin's latest DevFest Nantes 2019 talk he uses terms like instance rather than factory, and then it would be down to the agreed contract of the provided instance to know whether it was an object/factory function. So I suggest we go with terminology more like his latest talk or whatever terminology we find most clear.